File: sdks/typescript/trace-harness/src/loading/loader.ts
import { ConcurrencyPool } from "@compforge/harness-toolbox/concurrency";
import type { Source } from "../ingest/sources/base";
import type { TraceContext } from "../model/context";
import type { NormSpan } from "../model/span";
import { EvidenceMissing, EvidenceTooLarge, type LoadConfig } from "./model";
import { cloneSpan, EvidenceStore } from "./store";

type Fields = readonly string[] | null;
interface Request {
  trace: TraceContext; ids: readonly string[]; fields: Fields;
  resolve(): void; reject(error: unknown): void;
}
const covers = (span: NormSpan, fields: Fields): boolean => span.loaded_fields === null
  || fields !== null && fields.every(field => span.loaded_fields!.includes(field));
const union = (a: Fields, b: Fields): Fields => a === null || b === null ? null : [...new Set([...a, ...b])];

/** @spec Evidence is cached independently of computations; detail reads never change topology.
 * @why All leases of a trace coordinate through one loader, but receive separate mutable spans.
 */
export class EvidenceLoader {
  readonly stats = { reads: 0, fetches: 0, bytesRead: 0, cacheHits: 0 };
  readonly #pending = new Map<string, Request[]>();
  readonly #locks = new Map<string, Promise<unknown>>();
  constructor(readonly source: Source, readonly store: EvidenceStore,
    readonly config: Required<LoadConfig>, readonly pool: ConcurrencyPool, readonly signal: AbortSignal,
    readonly aliases: Readonly<Record<string, string>> = {},
    readonly normalize?: (span: NormSpan) => NormSpan) {}

  #serial<T>(key: string, work: () => Promise<T>): Promise<T> {
    const previous = this.#locks.get(key) ?? Promise.resolve();
    const next = previous.catch(() => {}).then(() => { this.signal.throwIfAborted(); return work(); });
    this.#locks.set(key, next);
    void next.finally(() => { if (this.#locks.get(key) === next) this.#locks.delete(key); }).catch(() => {});
    return next;
  }
  #fields(fields: Fields): Fields {
    return fields === null ? null : [...new Set([...fields,
      ...Object.entries(this.aliases).filter(([, canonical]) => fields.includes(canonical)).map(([alias]) => alias)])];
  }
  #normalized(span: NormSpan): NormSpan {
    const original = cloneSpan(span);
    for (const [alias, canonical] of Object.entries(this.aliases)) {
      if (!Object.hasOwn(original.attrs, canonical) && Object.hasOwn(original.attrs, alias)) {
        original.attrs[canonical] = original.attrs[alias];
      }
    }
    const result = this.normalize?.(original) ?? original;
    if (result.span_id !== span.span_id || result.parent_span_id !== span.parent_span_id) {
      throw new Error("normalization must preserve physical span identity and parent");
    }
    return result;
  }
  checkSize(spans: Map<string, NormSpan>): number {
    const size = [...spans.values()].reduce((sum, span) => sum + Buffer.byteLength(JSON.stringify(span)), 0);
    if (size > Math.min(this.config.maxTraceBytes, Math.floor(this.config.cacheBytes / this.config.activeTraces))) {
      throw new EvidenceTooLarge(`trace evidence requires ${size} bytes, exceeding loading budget`);
    }
    return size;
  }
  async skeleton(traceId: string, fields: readonly string[]): Promise<Map<string, NormSpan>> {
    return this.#serial(traceId, async () => {
      const key = `skeleton:${traceId}`;
      const saved = await this.store.get<Array<[string, NormSpan]>>(key);
      let spans: Map<string, NormSpan>;
      if (saved) {
        spans = new Map(saved.map(([id, span]) => [id, cloneSpan(span)]));
        // Extend projection through frozen identities, never fetch a new span set.
        await this.#read(traceId, spans, fields);
        this.stats.cacheHits++;
      } else {
        spans = await this.pool.run(() => this.source.fetch(traceId, this.#fields(fields)!, this.signal), this.signal);
        this.signal.throwIfAborted();
        if (!spans.size) throw new EvidenceMissing(`trace has no available spans: ${traceId}`);
        spans = new Map([...spans].map(([id, span]) => {
          if (id !== span.span_id) throw new Error(`invalid skeleton identity: ${traceId}/${id}`);
          const normalized = this.#normalized(span);
          normalized.loaded_fields = fields;
          return [id, normalized];
        }));
        this.stats.bytesRead += this.checkSize(spans);
        this.stats.fetches++;
      }
      this.checkSize(spans);
      await this.store.put(key, [...spans]);
      for (const [id, span] of spans) await this.#save(traceId, id, span);
      return spans;
    });
  }
  load(trace: TraceContext, ids: readonly string[], fields: Fields): Promise<void> {
    this.signal.throwIfAborted();
    for (const id of ids) if (!trace.spans.has(id)) throw new EvidenceMissing(`span absent from frozen skeleton: ${trace.trace_id}/${id}`);
    if (ids.every(id => covers(trace.spans.get(id)!, fields))) return Promise.resolve();
    return new Promise((resolve, reject) => {
      const requests = this.#pending.get(trace.trace_id) ?? [];
      requests.push({ trace, ids, fields, resolve, reject });
      this.#pending.set(trace.trace_id, requests);
      if (requests.length === 1) {
        // A microtask merges concurrent consumers without an arbitrary timer delay.
        queueMicrotask(() => {
          const batch = this.#pending.get(trace.trace_id)!;
          this.#pending.delete(trace.trace_id);
          void this.#serial(trace.trace_id, () => this.#flush(batch)).catch(error => {
            for (const request of batch) request.reject(error);
          });
        });
      }
    });
  }
  async #flush(requests: Request[]): Promise<void> {
    const fields = requests.reduce<Fields>((value, request) => union(value, request.fields), []);
    const spans = new Map<string, NormSpan>();
    for (const request of requests) for (const id of request.ids) spans.set(id, cloneSpan(request.trace.spans.get(id)!));
    try {
      await this.#read(requests[0]!.trace.trace_id, spans, fields);
      for (const request of requests) {
        const staged = new Map(request.trace.spans);
        for (const id of request.ids) staged.set(id, this.#merge(staged.get(id)!, spans.get(id)!, request.fields));
        this.checkSize(staged);
        this.signal.throwIfAborted();
        for (const id of request.ids) Object.assign(request.trace.spans.get(id)!, staged.get(id)!);
        request.resolve();
      }
    } catch (error) {
      for (const request of requests) {
        for (const id of request.ids) for (const name of request.fields ?? ["*"]) {
          request.trace.spans.get(id)!.field_errors[name] = error instanceof Error ? error.message : String(error);
        }
        request.reject(error);
      }
    }
  }
  #merge(target: NormSpan, incoming: NormSpan, fields: Fields): NormSpan {
    if (target.span_id !== incoming.span_id || target.parent_span_id !== incoming.parent_span_id
      || target.name !== incoming.name || target.start_ms !== incoming.start_ms || target.dur_ms !== incoming.dur_ms
      || target.service !== incoming.service || target.has_error !== incoming.has_error
      || target.storage_id !== incoming.storage_id || target.storage_index !== incoming.storage_index) {
      throw new EvidenceMissing(`evidence identity changed: ${target.span_id}`);
    }
    const merged = cloneSpan(target);
    const names = fields ?? Object.keys(incoming.attrs);
    for (const name of names) if (Object.hasOwn(incoming.attrs, name)) merged.attrs[name] = incoming.attrs[name];
    // Keep structural values from the skeleton. Only evidence payload is extended.
    // Raw documents are opaque Source snapshots. Do not merge backend-specific tag formats.
    // Once a full snapshot is cached, a later projection must not replace it with partial raw.
    merged.raw = target.loaded_fields === null && fields !== null ? merged.raw : structuredClone(incoming.raw);
    if (fields === null) { merged.events = incoming.events; merged.error_events = incoming.error_events; }
    merged.loaded_fields = union(target.loaded_fields, fields);
    for (const name of fields ?? ["*"]) delete merged.field_errors[name];
    return merged;
  }
  async #read(traceId: string, spans: Map<string, NormSpan>, fields: Fields): Promise<void> {
    const missing: string[] = [];
    for (const [id, span] of spans) {
      if (covers(span, fields)) continue;
      const cached = await this.store.get<NormSpan>(JSON.stringify(["span", traceId, id]));
      if (cached && covers(cached, fields)) { spans.set(id, this.#merge(span, cached, fields)); this.stats.cacheHits++; }
      else missing.push(id);
    }
    for (let offset = 0; offset < missing.length; offset += 100) {
      const ids = missing.slice(offset, offset + 100);
      const refs = ids.map(id => ({ trace_id: traceId, span_id: id,
        index: spans.get(id)!.storage_index, document_id: spans.get(id)!.storage_id }));
      const fetched = await this.pool.run(() => this.source.read(refs, this.#fields(fields), this.signal), this.signal);
      this.signal.throwIfAborted();
      this.stats.reads++;
      this.stats.bytesRead += this.checkSize(fetched);
      for (const id of ids) {
        const incoming = fetched.get(id);
        if (!incoming) throw new EvidenceMissing(`evidence expired or missing: ${traceId}/${id}`);
        spans.set(id, this.#merge(spans.get(id)!, this.#normalized(incoming), fields));
      }
    }
    this.checkSize(spans);
    // Failed reads never commit a partially prepared set as successful evidence.
    for (const id of missing) await this.#save(traceId, id, spans.get(id)!);
  }
  async #save(traceId: string, id: string, span: NormSpan): Promise<void> {
    const key = JSON.stringify(["span", traceId, id]);
    const cached = await this.store.get<NormSpan>(key);
    await this.store.put(key, cached ? this.#merge(cached, span, span.loaded_fields) : span);
  }
  async close(): Promise<void> {
    // Session aborts first, preventing queued work from reaching the Source.
    await Promise.resolve();
    await Promise.allSettled([...this.#locks.values()]);
  }
}

File: sdks/python/trace_harness/loading/loader.py
"""Coalesced, bounded evidence reads with durable cache reuse."""

from __future__ import annotations

import asyncio
import json
from dataclasses import asdict

from trace_harness.ingest.sources.base import Source
from trace_harness.loading.model import EvidenceMissing, EvidenceRef, EvidenceTooLarge, LoadConfig
from trace_harness.loading.store import EvidenceStore
from trace_harness.model.context import TraceContext
from trace_harness.model.span import NormSpan


class EvidenceLoader:
    def __init__(
        self,
        source: Source,
        store: EvidenceStore,
        config: LoadConfig,
        *,
        field_aliases=None,
        normalize_span=None,
    ):
        self.source, self.store, self.config = source, store, config
        self.field_aliases = field_aliases or {}
        self.normalize_span = normalize_span
        self._semaphore = asyncio.Semaphore(config.concurrency)
        self._locks = [asyncio.Lock() for _ in range(config.concurrency)]
        self._skeleton_locks = [asyncio.Lock() for _ in range(config.concurrency)]
        self._pending: dict[str, list] = {}
        self._tasks: set[asyncio.Task] = set()
        self.stats = {"cache_hits": 0, "reads": 0, "bytes_read": 0}
        self._closed = False

    async def skeleton(self, trace_id: str, fields: tuple[str, ...]):
        # A trace may have multiple concurrent leases. Freeze membership exactly once.
        async with self._skeleton_locks[hash(trace_id) % len(self._skeleton_locks)]:
            return await self._skeleton_locked(trace_id, fields)

    async def _skeleton_locked(self, trace_id, fields):
        key = f"skeleton:{trace_id}"
        saved = self.store.get(key)
        if saved is not None:
            spans = {sid: NormSpan(**value) for sid, value in saved["spans"].items()}
            extra = tuple(set(fields) - set(saved["fields"]))
            if extra:
                from trace_harness.model.context import TraceContext

                trace = TraceContext(trace_id, spans, [], {})
                await self.load(trace, tuple(spans), extra)
                self.store.put(
                    key,
                    {
                        "fields": sorted(set(fields) | set(saved["fields"])),
                        "spans": {sid: asdict(span) for sid, span in spans.items()},
                    },
                )
            return spans
        async with self._semaphore:
            spans = await self.source.fetch(trace_id, fields)
        if not spans:
            raise EvidenceMissing(f"trace has no available spans: {trace_id}")
        self._check_size(spans)
        for span in spans.values():
            span.loaded_fields = fields
        self.store.put(
            key, {"fields": fields, "spans": {sid: asdict(span) for sid, span in spans.items()}}
        )
        for span in spans.values():
            self.store.save(trace_id, span, fields)
        return spans

    def _check_size(self, spans):
        size = sum(len(json.dumps(s.raw).encode()) for s in spans.values())
        if size > min(
            self.config.max_trace_bytes, self.config.cache_bytes // self.config.active_traces
        ):
            raise EvidenceTooLarge(f"trace requires {size} bytes, exceeding loading budget")
        return size

    async def load(
        self, trace: TraceContext, span_ids: tuple[str, ...], fields: tuple[str, ...] | None
    ):
        if self._closed:
            raise RuntimeError("evidence loader is closed")
        if all(
            sid in trace.spans
            and (
                trace.spans[sid].loaded_fields is None
                or (fields is not None and set(fields).issubset(trace.spans[sid].loaded_fields))
            )
            for sid in span_ids
        ):
            return
        future = asyncio.get_running_loop().create_future()
        queue = self._pending.setdefault(trace.trace_id, [])
        queue.append((trace, span_ids, fields, future))
        if len(queue) == 1:
            task = asyncio.create_task(self._flush(trace.trace_id))
            self._tasks.add(task)
            task.add_done_callback(self._tasks.discard)
        await future

    async def _flush(self, trace_id):
        async with self._locks[hash(trace_id) % len(self._locks)]:
            await self._flush_locked(trace_id)

    async def _flush_locked(self, trace_id):
        # One event-loop turn collects concurrent requests without a latency timer.
        await asyncio.sleep(0)
        requests = self._pending.pop(trace_id, [])
        if not requests:
            return
        fields = (
            None
            if any(r[2] is None for r in requests)
            else tuple(sorted({f for r in requests for f in r[2]}))
        )
        spans = {
            sid: trace.spans[sid]
            for trace, ids, _, _ in requests
            for sid in ids
            if sid in trace.spans
        }
        try:
            unknown = {sid for _, ids, _, _ in requests for sid in ids} - spans.keys()
            if unknown:
                raise EvidenceMissing(f"span absent from frozen skeleton: {sorted(unknown)}")
            found, missing = {}, []
            for sid, span in spans.items():
                cached = self.store.read(trace_id, sid, fields)
                if cached is not None:
                    self.stats["cache_hits"] += 1
                    found[sid] = self.normalize_span(cached) if self.normalize_span else cached
                else:
                    missing.append(EvidenceRef(trace_id, sid, span.storage_index, span.storage_id))
            # Bounded batches avoid oversized bool queries and backend clause limits.
            for offset in range(0, len(missing), 100):
                refs = tuple(missing[offset : offset + 100])
                async with self._semaphore:
                    requested_fields = (
                        None
                        if fields is None
                        else tuple(
                            set(fields)
                            | {
                                alias
                                for alias, canonical in self.field_aliases.items()
                                if canonical in fields
                            }
                        )
                    )
                    fetched = await self.source.read(refs, requested_fields)
                self.stats["reads"] += 1
                self.stats["bytes_read"] += self._check_size(fetched)
                for sid, span in fetched.items():
                    if self.normalize_span:
                        span = self.normalize_span(span)
                    self.store.save(trace_id, span, fields)
                    found[sid] = span
            for trace, ids, requested, future in requests:
                if future.done():
                    continue
                absent = set(ids) - found.keys()
                if absent:
                    for sid in absent:
                        for name in requested or ("*",):
                            trace.spans[sid].field_errors[name] = "EvidenceMissing"
                    future.set_exception(
                        EvidenceMissing(f"evidence expired or missing: {trace_id}/{sorted(absent)}")
                    )
                    continue
                for sid in ids:
                    incoming, target = found[sid], trace.spans[sid]
                    attrs = (
                        incoming.attrs
                        if requested is None
                        else {k: v for k, v in incoming.attrs.items() if k in requested}
                    )
                    target.attrs.update(attrs)
                    tags = {t["key"]: t for t in target.raw.get("tags", [])}
                    tags.update(
                        {
                            t["key"]: t
                            for t in incoming.raw.get("tags", [])
                            if requested is None or t["key"] in requested
                        }
                    )
                    target.raw = {**incoming.raw, "tags": list(tags.values())}
                    target.loaded_fields = (
                        None
                        if requested is None or target.loaded_fields is None
                        else tuple(set(target.loaded_fields) | set(requested))
                    )
                    for name in requested or ("*",):
                        target.field_errors.pop(name, None)
                    if requested is None:
                        target.error_events, target.events = incoming.error_events, incoming.events
                self._check_size(trace.spans)
                future.set_result(None)
        except BaseException as exc:
            for trace, ids, fields, future in requests:
                for sid in ids:
                    if sid in trace.spans:
                        for name in fields or ("*",):
                            trace.spans[sid].field_errors[name] = type(exc).__name__
                if not future.done():
                    if isinstance(exc, asyncio.CancelledError):
                        future.cancel()
                    else:
                        future.set_exception(exc)

    async def close(self):
        self._closed = True
        for task in self._tasks:
            task.cancel()
        await asyncio.gather(*self._tasks, return_exceptions=True)
